rng/philox/philox.hpp

```diff
@@ -0,0 +1,298 @@
+    // types
+    using result_type = UIntType;
+
+    // engine characteristics
+    static constexpr std::size_t word_size = w;
+    static constexpr std::size_t word_count = n;
+    static constexpr std::size_t round_count = r;
+    static constexpr std::array<result_type, array_size> multipliers = extract_elements(even_indices_sequence{});
+    static constexpr std::array<result_type, array_size> round_consts = extract_elements(odd_indices_sequence{});
+    static constexpr result_type min() { return 0; }
+    static constexpr result_type max() { return max_impl(); }
+    static constexpr result_type default_seed = 20111115u;
+    // constructors and seeding functions
+    philox_engine() : philox_engine(default_seed) {}
+
+    explicit philox_engine(result_type value) : x{}, k{}, state_i{n - 1} {
+        k[0] = value & result_mask;
+    }
+
+    template<class Sseq>
+    explicit philox_engine(Sseq& q)
+    {
+        seed(q);
+    }
+
+    void seed(result_type value = default_seed) {
+        k[0] = value & result_mask;
+        for (std::size_t j = 1; j < array_size; ++j) {
+            k[j] = 0;
+        }
+        reset_counter();
+    }
+
+    template<class Sseq>
+    void seed(Sseq& q)
+    {
+        using seq_result_type = typename Sseq::result_type;
+
+        constexpr std::size_t p = (w - 1) / 32 + 1; // ceil division
+        std::array<seq_result_type, n / 2 * p> a;
+        q.generate(a.begin(), a.end());
+
+        for (std::size_t i = 0; i < (n / 2); ++i)
+        {
+            result_type sum = 0;
+            for (std::size_t j = 0; j < p; ++j)
+            {
+                sum += a[i * p + j] << (32 * j);
+            }
+            k[i] = sum & result_mask;
+        }
+
+        reset_counter();
+    }
+
+    void set_counter(const array<result_type, n>& counter) {
+        for (std::size_t j = 0; j < n; ++j) {
+            x[n - j - 1] = counter[j] & result_mask;
+        }
+    }
+
+    // equality operators
+    friend bool operator==(const philox_engine& left, const philox_engine& right) {
+        bool result = std::ranges::equal(left.x, right.x)
+            && std::ranges::equal(left.k, right.k)
+            && left.state_i == right.state_i;
+        for (auto i = left.state_i + 1; (i < n) && result; ++i)
+        {
+            result = result && (left.y[i] == right.y[i]);
+        }
+        return result;
+    }
+
+    // generating functions
+    result_type operator()()
+    {
+        ++state_i;
+        if (state_i == n) {
+            y = philox_generate(k, x); // see below
+            increment_counter();
+            state_i = 0;
+        }
+        return y[state_i];
+    }
+
+    void discard(unsigned long long z) {
+        std::uint32_t available_in_buffer = n - 1 - state_i;
+        if (z <= available_in_buffer) {
+            state_i += z;
+        }
+        else {
+            z -= available_in_buffer;
+            int tail = z % n;
+            if (tail == 0 && state_i == (n - 1)) {
```

Comment:
former state_i should not influence flow at this point.
should be
```
if(tail == 0) {
  state_i = n-1;
  increment_counter(z / n);
} ...
```

Reply:
Done